types/serve-static/index.d.ts

```diff
@@ -2,33 +2,28 @@
 // Project: https://github.com/expressjs/serve-static
 // Definitions by: Uros Smolnik <https://github.com/urossmolnik>
 //                 Linus Unnebäck <https://github.com/LinusU>
+//                 Devansh Jethmalani <https://github.com/devanshj>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.3
 
-/* =================== USAGE ===================
-
-    import * as serveStatic from "serve-static";
-    app.use(serveStatic("public/ftp", {"index": ["default.html", "default.htm"]}))
-
- =============================================== */
-
-/// <reference types="express-serve-static-core" />
-
-import * as express from "express-serve-static-core";
 import * as m from "mime";
+import * as http from "http";
 
 /**
  * Create a new middleware function to serve files from within a given root directory.
  * The file to serve will be determined by combining req.url with the provided root directory.
  * When a file is not found, instead of sending a 404 response, this module will instead call next() to move on to the next middleware, allowing for stacking and fall-backs.
  */
-declare function serveStatic(root: string, options?: serveStatic.ServeStaticOptions): express.Handler;
+declare function serveStatic<R extends http.OutgoingMessage>(
+    root: string,
+    options?: serveStatic.ServeStaticOptions<R>
+): serveStatic.RequestHandler<R>;
 
 declare namespace serveStatic {
     var mime: typeof m;
-    interface ServeStaticOptions {
+    interface ServeStaticOptions<R extends http.OutgoingMessage = http.OutgoingMessage> {
```

Comment:
Making the type parameter optional to avoid unnecessary breaking change for those who have used `ServeStaticOptions` in their codebase. I often avoid going so overboard to avoid breaking changes in types but it's not that bad hence went with it.

Reply:
Isn't this still a breaking change? Anyone who had previously used `serveStatic` with a custom `setHeaders` callback will now get the weaker `http.OutgoingMessage` instead of `express.Response` when *not* contextually typed. This is mitigated somewhat by the fact that `app.use(serveStatic(...))` should result in `R` being contextually typed to `express.Response`, but if anyone writes code like this then they will be broken:

```ts
const middleware = serveStatic(...); // no contextual type
app.use(middleware); // error: `OutgoingMessage` is not assignable to `Response`.
```

It would be good to have the tests verify the contextual type for the `app.use(serveStatic(...))` case. I'm not sure how often anyone writes `serveStatic(...)` outside of a `.use()`, so I'm not sure how much of an impact this change will be.

Reply:
Well ultimately it is a breaking change although making type parameter optional saves some error in cases where users have written `const options: ServeStaticOptions = { ... }` and are not using express exclusive methods on the response.

> [...] but if anyone writes code like this then they will be broken:
> ```
> const middleware = serveStatic(...); // no contextual type
> app.use(middleware); // error: `OutgoingMessage` is not assignable to `Response`.
> ```

Umm, not really? This would compile ...
```typescript
const middleware = serveStatic({ setHeaders: res => res.setHeader("foo, bar") }) // R is http.OutgoingMessage
app.use(middleware) // works because `use` will invoke with `express.Response` which is wider than `http.OutgoingMessage` which is required by `middleware`
```
... however this would not (which would have compiled with previous version) ...
```typescript
const middleware = serveStatic({ setHeaders: res => res.set("foo, bar") }) // R is http.OutgoingMessage, error: property set does not exist
app.use(middleware) // no error here
```

> It would be good to have the tests verify the contextual type for the app.use(serveStatic(...)) case.

Yes as you might have seen I have added tests for this and the above expected error as well.

> I'm not sure how often anyone writes serveStatic(...) outside of a .use(), so I'm not sure how much of an impact this change will be.

Yeah and even if they do write `serveStatic(...)` outside `use` it's won't be a breaking change as long as they don't use express exclusive methods in `setHeaders`. As I had said in the description:

> So it indeed is a breaking change but only for people who invoke the serveStatic with setHeaders option using express apis and the invocation is not directly passed to use (which makes the response not infer to express.Response).

But all things said I don't thing this will have a huge impact.